trieste/acquisition/function/function.py

```diff
@@ -583,6 +583,217 @@ def _update_expected_improvement_fn(
+        mean = tf.reduce_mean(samples_at_query_points, axis=0)
+
+        tf.debugging.assert_shapes(
+            [(mean, ["_", 1])], message="Expected model with output dimension [1]."
+        )
+
+        eta = tf.reduce_min(mean, axis=0)
+
+        return monte_carlo_expected_improvement(sampler, eta)
+
+
+class monte_carlo_expected_improvement(AcquisitionFunctionClass):
+    def __init__(self, sampler: ReparametrizationSampler[ProbabilisticModel], eta: TensorType):
+        r"""
+        Return a Monte Carlo based Expected Improvement (EI) acquisition function for
+        single-objective global optimization. Improvement is with respect to the current "best"
+        observation ``eta``, where an improvement moves towards the objective function's minimum
+        and the expectation is calculated with respect to the ``model`` posterior. For model
+        posterior :math:`f`, this is
+
+        .. math:: x \mapsto \mathbb E \left[ \max (\eta - f(x), 0) \right].
+
+        For the Monte Carlo version, the expectation is calculated by samples that we save.
+
+        :param sampler: The model sampler of the objective function.
+        :param eta: The "best" observation.
+        :return: The Monte Carlo expected improvement function. This function will raise
+            :exc:`ValueError` or :exc:`~tf.errors.InvalidArgumentError` if used with a batch size
+            greater than one.
+        """
+        self._sampler = sampler
+        self._eta = tf.Variable(eta)
+
+    @tf.function
+    def __call__(self, at: TensorType) -> TensorType:
+        tf.debugging.assert_shapes(
+            [(at, [..., 1, None])],
+            message="This acquisition function only supports batch sizes of one.",
+        )
+        samples = self._sampler.sample(tf.squeeze(at, -2))  # [S, N, 1]
+        improvement = tf.maximum(self._eta - samples, 0.0)  # [S, N, 1]
+        return tf.reduce_mean(improvement, axis=0)  # [N, 1]
+
+
+class MonteCarloAugmentedExpectedImprovement(
+    SingleModelAcquisitionBuilder[SupportsGetObservationNoise]
+):
+    """
+    Builder for a Monte Carlo-based augmented expected improvement function for use with a model
+    without analytical augmented expected improvement (e.g. a deep GP). The "best" value is taken to
+    be the minimum of the posterior mean at observed points.
+    """
+
+    def __init__(self, sample_size: int):
+        """
+        :param sample_size: The number of samples for each batch of points.
+        :raise tf.errors.InvalidArgumentError: If ``sample_size`` is not positive.
+        """
+        tf.debugging.assert_positive(sample_size)
+
+        super().__init__()
+
+        self._sample_size = sample_size
+
+    def __repr__(self) -> str:
+        """"""
+        return f"MonteCarloAugmentedExpectedImprovement({self._sample_size!r})"
+
+    def prepare_acquisition_function(
+        self,
+        model: SupportsGetObservationNoise,
```

Comment:
with the code above:
```suggestion
        model: SupportsReparamSamplerObservationNoise,
```